### PR TITLE
Csg imc solve

### PR DESCRIPTION
**Pull request by JensWehner**
implements https://github.com/votca/csg/issues/407

- [x] I do not know how to integrate this into the workflow now
 - [x] Some scripts will have to be deleted
 - [x] Merge after https://github.com/votca/csg/pull/410
 - [x] It is hard to test, because the data in the issue is numerically very unstable, which also explains why it likes to crash the tests.

### CODE REVIEW

**Comment by junghans** [comment]
I will fix the scripting, but let me upload more realistic input data first.

**Comment by JensWehner** [comment]
there was actually an error in my parser. So maybe the data is not so bad.

**Comment by junghans** [comment]
More realistic data: https://github.com/votca/csg/issues/407#issuecomment-533143017

**Comment by JensWehner** [comment]
I get exactly the same numbers, how can I write a test for this?

**Comment by JensWehner** [comment]
would be an integration test with your data.

**Comment by junghans** [comment]
@JensWehner see:
https://github.com/votca/csg/blob/master/src/tools/CMakeLists.txt#L115-L125
if you replace `csg_resample` by `csg_imc_solve` you are already half way there.

File: share/scripts/inverse/update_imc.sh [hunk 1]
```diff
@@ -38,7 +37,7 @@ for group in $imc_groups; do
   # currently this is a hack! need to create combined array
   msg "solving linear equations for imc group '$group'"
   critical csg_imcrepack --in ${group} --out ${group}.packed
-  do_external imcsolver $solver  ${group}.packed ${group}.packed.sol $reg
+  critical csg_imc_solve --imcfile "${group}.packed.imc" --gmcfile "${group}.packed.gmc" --regularization "${reg}" --outputfile "${group}.packed.sol"
   critical csg_imcrepack --in ${group}.packed --unpack ${group}.packed.sol
```
Comment (junghans) [inline share/scripts/inverse/update_imc.sh:41]: @JensWehner, can be combine this series of `csg_imcrepack`, `csg_imc_solve`, `csg_imcrepack` into one thing?

Comment (junghans) [reply share/scripts/inverse/update_imc.sh:41]: I put the input files to `csg_imcrepack` here: https://github.com/votca/csg/issues/407#issuecomment-533553172

Comment (JensWehner) [reply share/scripts/inverse/update_imc.sh:41]: can you add the final output as well?

Comment (junghans) [reply share/scripts/inverse/update_imc.sh:41]: Without the `--unpack` option, it generates the `CG-CG.packed` files, with the `--unpack` option is generates the potentials e.g. *.imc.dpot`.

I uploaded a complete set here:
https://github.com/votca/csg/issues/407#issuecomment-533628560

**Comment by JensWehner** [comment]
I have no really good idea, but I think considering that repack and csg_imc_solve do not work on the same things. I would do the merge of imc_repack and csg_imc_solve at a later stag.

**Comment by junghans** [comment]
Sounds good to me.

**Comment by JensWehner** [comment]
some scripts stuff still has to be removed correct?

**Comment by junghans** [comment]
Done, forgot to push it.

**Comment by junghans** [comment]
One last question, `csg_imc_solve` vs. `csg_imcsolve` (as in similar to `csg_imcrepack`), what do you like better as a name?

Also @DRosen285 wanted to test it, too.

**Comment by JensWehner** [comment]
I like the one with two underscores a bit more, but consistency is also nice. So I have no strong opinion.

**Comment by junghans** [comment]
@JensWehner ok then lets leave it like it is.

**Comment by JensWehner** [comment]
@junghans you have to review and you can merge then from my side.

**Comment by junghans** [comment]
Fine with me, let's give @DRosen285 another day to look at it.

**Comment by DRosen285** [comment]
Sorry for the delay. 
Looks good to me. 
Just one comment following the discussion under #407 : it actually makes more sense to have one regularization parameter per group, instead per interaction. As the regularization term acts on the whole matrix, if multiple interactions are in the same group.

And one question just for curiosity: How do you determine the numerical stability of the IMC scheme? 
In the past I've detected it only based on how well the updated potentials perform in the next iteration. But it seems you keep track of the stability on the fly in csg_imc_solve.cc

**Comment by JensWehner** [comment]
@DRosen285 I do not fully understand, but concerning the numerical instability. 

The problematic part of solving the problem is calculating the inverse of (A.T*A), because if the eigenvalues of (A.T*A) are very small you will nearly divide by zero. Small eigenvalues are normally a symptom of bad input data. So there is very little you can do. 

- Use regularisation, as you basicallly add a constant to each eigenvalue. 
- Calculate not the inverse of A.T*A, but the pseudo inverse. https://en.wikipedia.org/wiki/Generalized_inverse , basically you remove eigenvectors with very small eigenvalues (<epsilon) from the calculation. Depending on the epsilon, this will stabilize the numerics of your calculation.

**Comment by DRosen285** [comment]
@JensWehner got it. I just got confused, because I used the pseudoinverse in combination with regularisation in the past. Because as I understand it, the pseudoinverse only helps to make a matrix invertible in case it is not. But one might still need regularization, because also for the pseudoinverse small eigenvalues might still exist (at least that is what I oberseved in the past).

**Review by junghans — APPROVED** [review]
LGTM

**Comment by junghans** [comment]
I will fix #413 in a separate PR.

**Pull request merged**